File: programmers/프로그래머스_합승택시.cpp
#include <string>
#include <vector>
#include <queue>

#define MAX 2000000000
using namespace std;
void Dijkstra(int start,int dp[201],vector<pair<int,int> > v[201]){
    priority_queue< pair<int,int> ,vector<pair<int,int>>,greater<pair<int,int>> > pq;
    int curNode,curCost;
    pair<int,int> temp;
    dp[start]=0;
    pq.push({0,start});
    while(!pq.empty()){
        temp=pq.top();
        pq.pop();
        curCost=temp.first,curNode=temp.second;
        if(curCost>dp[curNode]) continue;
        for(int i=0;i<v[curNode].size();i++){
            int candidateNode=v[curNode][i].first;
            int candidateNodeCost=v[curNode][i].second;
            if(curCost+candidateNodeCost<dp[candidateNode]){
                dp[candidateNode]=curCost+candidateNodeCost;
                pq.push({dp[candidateNode],candidateNode});
            }
        }
    }
}
int solution(int n, int s, int a, int b, vector<vector<int>> fares) {
    int answer = MAX;
    int startNode,endNode,cost;
    int dp[3][201];
    for(int j=0;j<3;j++)
        for(int i=0;i<201;i++) dp[j][i]=MAX;
    vector<pair<int,int> > v[201];
    for(int i=0;i<fares.size();i++){
        startNode=fares[i][0];
        endNode=fares[i][1];
        cost=fares[i][2];
        v[startNode].push_back({endNode,cost});
        v[endNode].push_back({startNode,cost});
    }
    Dijkstra(s,dp[0],v);
    Dijkstra(a,dp[1],v);
    Dijkstra(b,dp[2],v);
    answer=dp[0][a]+dp[0][b];
    for(int i=1;i<=n;i++) answer=min(answer,dp[0][i]+dp[1][i]+dp[2][i]);
    return answer;
}

File: programmers/프로그래머스_합승택시.py
import heapq
MAX=2000000000
def Dijkstra(start:int,end:int,n:int,v:list)->int:
    pq=[[0,start]]
    dp=[MAX for i in range(n+1)]
    dp[start]=0
    while pq:
        curCost,curNode=heapq.heappop(pq)
        if curCost>dp[curNode]:
            continue
        for i in v[curNode]:
            nextNode,nextNodeCost=i
            if dp[nextNode]>curCost+nextNodeCost:
                dp[nextNode]=curCost+nextNodeCost
                heapq.heappush(pq,[curCost+nextNodeCost,nextNode])
    return dp[end]
    pass
def solution(n, s, a, b, fares):
    answer = 0
    v=[[] for i in range(n+1)]
    for i in  fares:
        start,end,cost=i
        v[start].append([end,cost])
        v[end].append([start,cost])
    answer=Dijkstra(s,a,n,v)+Dijkstra(s,b,n,v)
    for i in range(1,n+1):
        if i!=s:
            answer=min(Dijkstra(s,i,n,v)+Dijkstra(i,b,n,v)+Dijkstra(i,a,n,v),answer) 
    return answer
